Add a `try_collect()` helper method to `Iterator`

a-lafrance opened the pull request:
Implement `Iterator::try_collect()` as a helper around `Iterator::collect()` as discussed [here](https://internals.rust-lang.org/t/idea-fallible-iterator-mapping-with-try-map/15715/5?u=a.lafrance).

First time contributor so definitely open to any feedback about my implementation! Specifically wondering if I should open a tracking issue for the unstable feature I introduced.

As the main participant in the internals discussion: r? @scottmcm

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @scottmcm (or someone else) soon.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

scottmcm commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -1777,6 +1777,62 @@ pub trait Iterator {
+        Self: Sized,
+        <Self as Iterator>::Item: Try,
+        <<Self as Iterator>::Item as Try>::Residual: Residual<T>,
+        ChangeOutputType<Self::Item, T>: FromIterator<Self::Item>,
```
Hmm, personally I wasn't thinking of having this go through `FromIterator` for `Result` at all, since that would mean that people still need to *write* that impl.

If this were to go through `iter::try_process` instead, for example, then it wouldn't need `Result: FromIterator`

https://github.com/rust-lang/rust/blob/393fdc10483da930cdbb00eabc3635030d2e776f/library/core/src/iter/adapters/mod.rs#L140-L156

(That might not work exactly as-is, though.  It might need a new `try_process_mut(&mut I, F)` version so that continuing after starting works out well, which might mean changing `GenericShunt` to also hold `&mut 'a I` instead of just `I`.)

But regardless of the precise implementation details, I'd suggest trying to get this to work:
```rust
    use std::ops::ControlFlow::{Break, Continue};
    let a = [Continue(1), Continue(2), Break(3), Continue(4), Continue(5)];
    let mut it = a.into_iter();
    let r = it.try_collect::<Vec<_>>();
    assert_eq!(r, Break(3));
    let r = it.try_collect::<Vec<_>>();
    assert_eq!(r, Continue(vec![4, 5]));
```
It might need some more type hints in the asserts, but otherwise it should work, I think.  And it hits the core things here: a `Try`-but-not-`FromIterator` type, the resumption behaviour, and the turbofishing being only the collection.

a-lafrance replied:
Ok yeah, I see what you're saying. I think I have a better grasp of the main points of the feature now, so I'll modify my implementation per your comments. Thanks!

scottmcm commented on library/core/src/iter/traits/iterator.rs:
```diff
@@ -1777,6 +1777,62 @@ pub trait Iterator {
+    ///
+    /// [`collect`]: Iterator::collect
+    #[inline]
+    #[unstable(feature = "iterator_try_collect", issue = "none")]
```
And yes, you'll open a "Library Tracking Issue" (it's one of the templates in https://github.com/rust-lang/rust/issues/new/choose ) at some point to put here.  But you don't need to do that until it's going to go in, so it's often reasonable to not bother doing so until you get a "sure, we can put something like this in nightly" from a reviewer.

Sometimes people are confident and open them along with the PR; sometimes the PRs are more speculative and they wait.  It's up to you.

a-lafrance replied:
Alright, I'll probably hold off on opening anything until I've got a reasonable implementation then

scottmcm replied:
Oh, just `try_process(self, |i| i.collect())` worked?  That's convenient 🚀 

I think this looks good.  Can you open the tracking issue and update the attribute?  (And maybe squash the commits too, for cleanliness.)

Then once the check build is fixed -- I don't see how it could be this PR's fault -- I'll approve it.

a-lafrance replied:
Sounds good! And yeah, I was impressed that it worked right off the bat too 😆

a-lafrance commented:
Updated the `try_collect()` implementation to accept `Try`-not-`FromIterator` types like `ControlFlow`. Also modified the docs to get the main points of the method across better.

@rustbot label -S-waiting-on-author +S-waiting-on-review

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5211800302?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/619de2156f2f83a9a810850aaebe5d49c171e8c7/checks/5211800302/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
 Documenting std v0.0.0 (/checkout/library/std)
    Finished release [optimized] target(s) in 10.10s
    Checking std v0.0.0 (/checkout/library/std)
 Documenting proc_macro v0.0.0 (/checkout/library/proc_macro)
error[E0773]: attempted to define built-in macro more than once
     |
1306 | /     macro_rules! cfg {
1306 | /     macro_rules! cfg {
1307 | |         ($($cfg:tt)*) => {
1309 | |         };
1310 | |     }
     | |_____^
     |
     |
note: previously defined here
    --> /checkout/library/core/src/macros/mod.rs:1306:5
     |
1306 | /     macro_rules! cfg {
1307 | |         ($($cfg:tt)*) => {
1309 | |         };
1310 | |     }
     | |_____^

---
For more information about this error, try `rustc --explain E0773`.
error: could not document `proc_macro`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name proc_macro library/proc_macro/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc --error-format=json --json=diagnostic-rendered-ansi,future-incompat --markdown-css rust.css --markdown-no-toc -Z unstable-options --resource-suffix 1.60.0 --index-page /checkout/src/doc/index.md -C metadata=030f61473e41e060 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --extern std=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libstd-8ecdf0b3fdf1770d.rmeta --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.60.0-nightly
  (29f6d4c13
  2022-02-16)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")'` (exit status: 1)
```

</details>

scottmcm commented:
That CI failures makes no sense.  I'll try closing and reopening.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/5212509655?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/619de2156f2f83a9a810850aaebe5d49c171e8c7/checks/5212509655/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
 Documenting std v0.0.0 (/checkout/library/std)
    Finished release [optimized] target(s) in 8.67s
    Checking std v0.0.0 (/checkout/library/std)
 Documenting proc_macro v0.0.0 (/checkout/library/proc_macro)
error[E0773]: attempted to define built-in macro more than once
     |
1306 | /     macro_rules! cfg {
1306 | /     macro_rules! cfg {
1307 | |         ($($cfg:tt)*) => {
1309 | |         };
1310 | |     }
     | |_____^
     |
     |
note: previously defined here
    --> /checkout/library/core/src/macros/mod.rs:1306:5
     |
1306 | /     macro_rules! cfg {
1307 | |         ($($cfg:tt)*) => {
1309 | |         };
1310 | |     }
     | |_____^

---
For more information about this error, try `rustc --explain E0773`.
error: could not document `proc_macro`

Caused by:
  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustdoc --edition=2021 --crate-type lib --crate-name proc_macro library/proc_macro/src/lib.rs --target x86_64-unknown-linux-gnu -o /checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/doc --error-format=json --json=diagnostic-rendered-ansi,future-incompat --markdown-css rust.css --markdown-no-toc -Z unstable-options --resource-suffix 1.60.0 --index-page /checkout/src/doc/index.md -C metadata=030f61473e41e060 -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --extern std=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libstd-8ecdf0b3fdf1770d.rmeta --cfg=bootstrap -Csymbol-mangling-version=legacy -Zunstable-options -Dwarnings '-Wrustdoc::invalid_codeblock_attributes' --crate-version '1.60.0-nightly
  (8cf0dcde1
  2022-02-16)' '-Zcrate-attr=doc(html_root_url="https://doc.rust-lang.org/nightly/")'` (exit status: 1)
```

</details>

a-lafrance commented:
Tracking issue: #94047

bjorn3 commented on library/core/tests/iter/traits/iterator.rs:
```diff
@@ -497,6 +497,52 @@ fn test_collect() {
+
+    let u: Vec<Result<i32, ()>> = vec![Ok(1), Ok(2), Ok(3)];
+    let v = u.into_iter().try_collect::<Vec<i32>>();
+    assert_eq!(v, Ok(vec![1, 2, 3]));
```
This is already possible with `.collect()`. Thanks to `impl<A, E, V> FromIterator<Result<A, E>> for Result<V, E> where V: FromIterator<A>`.

a-lafrance replied:
Right, but the main benefits of `try_collect()` are increasing the discoverability of this technique and allowing `Try`-but-not-`FromIterator` types (e.g. `ControlFlow`) to be fallibly collected as well. See my comment on the tracking issue [here](https://github.com/rust-lang/rust/issues/94047#issuecomment-1042094358) for a more thorough description of the differences between `try_collect()` and `collect()`.

ehuss commented:
The ICE in CI is caused by the use of intra-doc links involving `core` (the ones that are ``` [`ControlFlow`]: core::ops::ControlFlow` ``` and ``` [`FromIterator`]: core::iter::FromIterator ```).  There is more information in #73445. Obviously it shouldn't ICE and fall apart. I can't explain why that happens.

I'm pretty sure you can just reference them without qualification (like ```[`FromIterator`]```, no need for a reference) since they are in scope.

You can test documentation with `./x.py doc --stage 0 library/test`.  You can add the `--open` flag to open a web browser to inspect it.

a-lafrance commented:
Thanks for the info @ehuss! Just pushed a fix for the erring links, so @scottmcm I think it should be ready for a final review.

@rustbot label -S-waiting-on-author +S-waiting-on-review

a-lafrance commented:
Oh, and @rustbot label -S-blocked

scottmcm commented:
Thank you, ehuss!

This looks good to go into unstable by me.  Lots more discussion to be had on the tracking issue, to be sure.
@bors r+

bors commented:
:pushpin: Commit 47d5196a00e25ef5683fd58f472ce44c82f8ae06 has been approved by `scottmcm`

<!-- @bors r=scottmcm 47d5196a00e25ef5683fd58f472ce44c82f8ae06 -->
<!-- homu: {"type":"Approved","sha":"47d5196a00e25ef5683fd58f472ce44c82f8ae06","approver":"scottmcm"} -->
